File: 386-lexicographical-numbers/lexicographical-numbers.py
# Brute Force Code & Optimal Code
class Solution:
    # recursive function to generate numbers in lexicographical order
    def solve(self, current: int, n: int, result: list[int]) -> None:
        # stop if the current number exceeds n
        if current > n:
            return
        # add the current number to the result
        result.append(current)
        # try appending digits from 0 to 9
        for next_digit in range(10):
            next_num = current * 10 + next_digit
            # no need to continue if the number exceeds n
            if next_num > n:
                return
            # recur for the next lexicographical number
            self.solve(next_num, n, result)
    # return numbers from 1 to n in lexicographical order
    def lexicalOrder(self, n: int) -> list[int]:
        result = []
        # start DFS from each leading digit (1 to 9)
        for num in range(1, 10):
            self.solve(num, n, result)
        return result

# Time Complexity : O(N)
# Space Complexity : O(N)

File: 386-lexicographical-numbers/lexicographical-numbers.cpp
// Brute Force Code & Optimal Code
class Solution {
public:
    // recursive function to generate numbers in lexicographical order
    void solve(int current, int n, vector<int>& result) {
        // stop if the current number exceeds n
        if (current > n) {
            return;
        }
        // add the current number to the result
        result.push_back(current);
        // try appending digits from 0 to 9
        for (int next_digit = 0; next_digit <= 9; next_digit++) {
            // create the next number by appending next_digit to current
            int next_num = current * 10 + next_digit;
            // no need to continue if the number exceeds n
            if (next_num > n) {
                return;
            }
            // recursively generate the next lexicographical number
            solve(next_num, n, result);
        }
    }
    // return numbers from 1 to n in lexicographical order
    vector<int> lexicalOrder(int n) {
        // store the final lexicographical order
        vector<int> result;
        // start DFS from every leading digit from 1 to 9
        for (int num = 1; num <= 9; num++) {
            solve(num, n, result);
        }
        // return the result
        return result;
    }
};

// Time Complexity : O(N)
// Space Complexity : O(N)